osquery/utils/aws_util.cpp

```diff
@@ -321,6 +323,52 @@ void initAwsSdk() {
   }
 }
 
+void getInstanceIDAndRegion(std::string& instance_id, std::string& region) {
+  static std::atomic<bool> checked(false);
+  static std::string cached_id;
+  static std::string cached_region;
+  if (checked) {
+    // Return if already checked
+    instance_id = cached_id;
+    region = cached_region;
+    return;
+  }
+
+  static std::once_flag once_flag;
+  std::call_once(once_flag, []() {
+    if (checked) {
+      return;
+    }
+
+    initAwsSdk();
+    http::client::request req(
+        "http://169.254.169.254/latest/dynamic/instance-identity/document");
+    http::client::options options;
+    options.timeout(3);
+    http::client client(options);
+
+    try {
+      http::client::response res = client.get(req);
+      if (res.status() == 200) {
+        pt::ptree tree;
+        std::stringstream ss(res.body());
+        pt::read_json(ss, tree);
+        cached_id = tree.get<std::string>("instanceId", ""),
+        cached_region = tree.get<std::string>("region", ""),
+        VLOG(1) << "EC2 instance ID: " << cached_id
+                << ". Region: " << cached_region;
+      }
+    } catch (const std::system_error& e) {
+      // Assume that this is not EC2 instance
+      VLOG(1) << "Error getting EC2 instance information: " << e.what();
```

Comment:
Assume this happens, then checked is set to true. We then continue and set `instance_id` and `region_id` to uninitialized values. We also will always return uninitialized values because of the caching.